docs/man/gchem3d-viewer.1
.\"     Title: gchem3d-viewer
.\"    Author: Jean Br\('efort <jean.brefort@normalesup.org>
.\" Generator: DocBook XSL Stylesheets v1.72.0 <http://docbook.sf.net/>
.\"      Date: $Date: 2007-05-07 19:33:06 $
.\"    Manual: gnome-chemistry-utils
.\"    Source: gcu 0.7.95
.\"
.TH "GCHEM3D\-VIEWER" "1" "$Date: 2007-05-07 19:33:06 $" "gcu 0.7.95" "gnome\-chemistry\-utils"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
gchem3d\-viewer \- a small chemical viewer application
.SH "SYNOPSIS"
.HP 15
\fBgchem3d\-viewer\fR [\fBOPTION(S)\fR...] [\fIFILE\fR]
.SH "DESCRIPTION"
.PP
\fBgchem3d\-viewer\fR
is a small chemical viewer application, which can show several chemical file formats.
.SH "OPTIONS"
.PP
The following options are accepted:
.PP
\fB\-b \fR\fB\fICOLOR\fR\fR, \fB\-\-bgcolor=\fR\fB\fICOLOR\fR\fR
.RS 4
Use the given color as background color.
\fICOLOR\fR
can be one of
"\fIblack\fR" (default), "\fIwhite\fR", "\fI#rrggbb\fR" (don't forget to escape the \(lq#\(rq character in the shell).
.RE
.PP
\fB\-d \fR\fB\fIMODEL\fR\fR, \fB\-\-display3d=\fR\fB\fIMODEL\fR\fR
.RS 4
Choose how molecules are displayed.
\fIMODEL\fR
can be one of
"\fIBallnStick\fR" (default), "\fISpaceFill\fR".
.RE
.PP
\fB\-?\fR, \fB\-\-help\fR
.RS 4
Show application help options.
.RE
.PP
\fB\-\-help\-all\fR, \fB\-\-help\-*\fR
.RS 4
Print all or just a group of help options. These options are not documented here. Instead see
\fBgtk\-options\fR(7)
and
\fBgnome\-options\fR(7).
.RE
.PP
\fB\-v\fR, \fB\-\-version\fR
.RS 4
Print gchem3d\-viewer version information.
.RE
.SH "SEE ALSO"
.PP
\fBgnome\-options\fR(7),
\fBgtk\-options\fR(7)
.SH "AUTHORS"
.PP
\fBJean Br\('efort\fR <\&jean.brefort@normalesup.org\&>
.sp -1n
.IP "" 4
Program author.
.PP
\fBDaniel Leidert\fR <\&daniel.leidert@wgdd.de\&>
.sp -1n
.IP "" 4
Manpage author.
.SH "COPYRIGHT"
Copyright \(co 2002\-2007 Jean Br\('efort
.br
Copyright \(co 2004\-2007 Daniel Leidert
.br
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 any later version published by the Free Software Foundation.
.br


NEWS
Version 0.7.96
	* #19835: gchempaint 0.7.95 still looks for libgoffice-0.3.pc.
	* #19836: Add KDE MIME stuff for gnome-crystal.
	* removed a localized string from gcrystal.schemas.in.

Version 0.7.95
	* uses goffice-0.4.
	* added a man page for gcrystal (Daniel Leidert).
	* cleaned build system (Daniel Leidert).
	* various minor fixes.

Version 0.7.91
	* fixes some internationalization issues.
	* #19523: fixes build issues.

Version 0.7.90
	* removed unstable stuff.

Version 0.7.5
	periodic table:
		* add a widget to control family color scheme (Jordan Mantha).
	Library:
		* add a new DialogOwner class and make the Application and
		Document classes derive from it.
	Other:
		* add or update API and applications documentation.
		* fixed various issues in both the library and the applications.

Version 0.7.4
	3d viewer:
		* add recent files support.
		* image export at any resolution.
	Gnome Crystal:
		* add recent files support.
		* image export at any resolution.
		* share more code with the 3d molecules viewer.
	periodic table:
		* add new color schemes to the table (thanks to Jordan Mantha
		for his help with this feature).
	Library:
		* extend color schemes mechanism in the GtkPeriodic widget.
		* now parse all BODR data in element class.
	Other:
		* fixed various build issues (Daniel Leidert).

Version 0.7.3
	3d viewer: open several documents at once, and use new 3d classes
	in the library..
	Gnome Crystal: use the new 3d classes and macros in the library.
	periodic table: display melting and boiling points (needs bodr-4).
	Library: some new classes used for 3d display in both Gnome Crystal
	and the 3d viewer; also some new macros related to get/set class
	properties and GConf entries.

Version 0.7.2
	Fix a 32 bits compilation issue.

Version 0.7.1
	Gnome Crystal: changed the user interface for a more HIG compilant one.
	Library: various little changes to the gcu::Application and
	gcu::FileChooser classes.
	Database: fixed bromine molar weight.

Version 0.7.0
	fixed --version command line option.
	fixed some build issues.
	include Gnome Crystal.

Version 0.5.4
	periodic table: display atomic radii, copy charts to other apps.
	mozilla plugin and 2d viewer: support background changes
	and spacefill model display.
	gchemcalc: copy isotopic patterns charts to other apps.
	
Version 0.5.3
	periodic table: display ionization energy and electron affinity.
	mozilla plugin: display pdb and mol files, and support limited
	bgcolor and display3d properties.
	gchemcalc: accepts a formula from the command line.
	3d viewer: new menus entries and command line options for
	background color and display mode.
	GtkComboPeriodic: new periodic table combo box.
	Some minor bugs fixed.
	New translation: ru (thanks to Valek Filippov).

Version 0.5.2
	Correct mean molecular mass for isotopic patterns of macromolecules.
	Add a periodic table application.
	Add a mozilla plugin skeleton.

Version 0.5.1
	Add mass composition and isotopic patterns to Calculator.

Version 0.5.0
	New Formula class.
	New Calculator program.
	Enhanced 3d molecular viewer.
	The database has been split.
	Various bugs fixed and api enhncements..

Version 0.3.2
	Several bugs fixed, essentially in Object class.
	Changed signature of Object::OnSignal. Leads to a compatibility break.

Version 0.3.1
	Object class has been extended with new functionalities (signals and links), and
	Object contextual menu now uses GtkAction based menus.
	Several bugs and memory leaks have been fixed.
	Support has been added for shared mime info but
	mime types tree will be polluted with chemical/* types.

Version 0.3.0
	The libraries have been merged into one.
	A new Document class.
	The Object class has been enriched with a hierarchical rules api (for example a molecule
	must contain at leat one atom and an atom cannot contain a molecule).
	Several bugs fixed.
	Default colors for some elements have changed to be compatible with a new
	coloring scheme common to several open source projects.

Version 0.1.7
	Added a maximum number of bonds for atoms.
	Added some documentation.
	Some code cleaning in crystalviewer directory.
	Some bugs fixed.

Version 0.1.6
	Added bgcolor property to GtkChem3DViewer
	Several bugs fixed.

Version 0.1.5
	Many bugs fixed in the radii database stuff
	Change of coordinates in GtkChem3DViewer

Version 0.1.4
	New chem3d widget and bonobo server do display 3d molecular structures
	Atomic radii and electronegativities added to the elements database

Version 0.1.3
	Many API changes, see ChangeLog.
	
Version 0.1.2
	Minor changes to the Object and Atom classes.
	
Version 0.1.1
	All APIs has been thoroughly revised.

Version 0.1.0
	First release of Gnome Chemistry Utils


programs/gchem3d-viewer.c
/* 
 * Gnome Chemistry Utils
 * programs/gchem3d-viewer.c
 *
 * Copyright (C) 2004-2006 Jean Bréfort <jean.brefort@normalesup.org>
 *
 * This program is free software; you can redistribute it and/or 
 * modify it under the terms of the GNU General Public License as 
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
 * USA
 */

#include "config.h"
#include <gcu/chemistry.h>
#include <gcu/gtkchem3dviewer.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <stdio.h>
#include <string.h>
#include <goffice/goffice.h>
#include <goffice/gtk/go-action-combo-color.h>
#include <libgnomevfs/gnome-vfs.h>
#include <libgnomeprint/gnome-print.h>
#include <libgnomeprint/gnome-print-job.h>
#include <libgnomeprintui/gnome-print-dialog.h>
#include <libgnomeprintui/gnome-print-job-preview.h>

/*!\file
A simple sample of the use of the GtkChem3DViewer widget.
*/

static void on_file_open (GtkWidget *widget, void *data)
{
	char* filename;
	GtkFileChooserDialog *dialog = (GtkFileChooserDialog*) gtk_file_chooser_dialog_new (
															_("Open"),
															GTK_WINDOW (data),
															GTK_FILE_CHOOSER_ACTION_OPEN,
															GTK_STOCK_OPEN, GTK_RESPONSE_ACCEPT,
															GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
															NULL);
	GtkFileChooser* chooser = GTK_FILE_CHOOSER (dialog);
	GtkFileFilter* filter = gtk_file_filter_new ();
	GtkChem3DViewer *viewer = GTK_CHEM3D_VIEWER (g_object_get_data (G_OBJECT (data), "viewer"));
	gtk_dialog_set_default_response (GTK_DIALOG (dialog), GTK_RESPONSE_OK);
	gtk_file_filter_add_mime_type (filter, "chemical/x-mdl-molfile");
	gtk_file_filter_add_mime_type (filter, "chemical/x-pdb");
	gtk_file_filter_add_pattern (filter, "*.pdb"); /* I don't know why mime type is not enough */
	gtk_file_filter_add_mime_type (filter, "chemical/x-xyz");
	gtk_file_chooser_set_filter (chooser, filter);
	gtk_file_chooser_set_local_only (chooser, FALSE);
	while (gtk_dialog_run (GTK_DIALOG (dialog)) == GTK_RESPONSE_ACCEPT)
	{
		filename = gtk_file_chooser_get_uri (chooser);
		gtk_chem3d_viewer_set_uri (viewer, filename);
		break;
	}
	gtk_widget_destroy (GTK_WIDGET (dialog));
}

static void on_file_print (GtkWidget *widget, void *data)
{
	GnomePrintConfig* config = gnome_print_config_default ();
	GnomePrintContext *pc;
	GnomePrintJob *gpj = gnome_print_job_new (config);
	int do_preview = 0, copies = 1, collate = 0;
	GnomePrintDialog *gpd;
	GtkChem3DViewer *viewer = GTK_CHEM3D_VIEWER (g_object_get_data (G_OBJECT (data), "viewer"));
	gpd = GNOME_PRINT_DIALOG (gnome_print_dialog_new (gpj, (const guchar*) "Print test", GNOME_PRINT_DIALOG_COPIES));
	gnome_print_dialog_set_copies (gpd, copies, collate);
	switch (gtk_dialog_run (GTK_DIALOG (gpd)))
	{
	case GNOME_PRINT_DIALOG_RESPONSE_PREVIEW:
		do_preview = 1;
		break;
	case GNOME_PRINT_DIALOG_RESPONSE_CANCEL:
		gtk_widget_destroy (GTK_WIDGET (gpd));
		g_object_unref (gpj);
		gnome_print_config_unref (config);
		return;
	}
	gtk_widget_destroy (GTK_WIDGET (gpd));
	pc = gnome_print_job_get_context (gpj);
	gnome_print_beginpage (pc, (const guchar*)"");
	gdouble width, height;
	gnome_print_config_get_double (config, (const guchar*) GNOME_PRINT_KEY_PAPER_WIDTH, &width);
	gnome_print_config_get_double (config, (const guchar*) GNOME_PRINT_KEY_PAPER_HEIGHT, &height);
	gtk_chem3d_viewer_print (viewer, pc, width, height);
	gnome_print_showpage (pc);
	g_object_unref (pc);
	gnome_print_job_close (gpj);
	if (do_preview)
	{
		GtkWidget *preview = gnome_print_job_preview_new (gpj, (const guchar*) _("Preview"));
		gtk_widget_show (preview);
	} else {
		gnome_print_job_print (gpj);
	}
	g_object_unref (gpj);
	gnome_print_config_unref (config);
}

static void on_quit (GtkWidget *widget, void *data)
{
	gtk_main_quit();
}

static void on_color_changed (GOActionComboColor *combo, GObject *obj)
{
	GOColor color = go_action_combo_color_get_color (combo, FALSE);
	char *scolor = g_strdup_printf ("#%02x%02x%02x", UINT_RGBA_R (color),
		UINT_RGBA_G (color), UINT_RGBA_B (color));
	g_object_set (obj, "bgcolor", scolor, NULL);
	g_free (scolor);
}

static void on_display (GtkRadioAction *action, GtkRadioAction *current, GObject *obj)
{
	g_object_set (obj, "display3d", gtk_radio_action_get_current_value (action), NULL);
}

static void on_about_activate_url (GtkAboutDialog *about, const gchar *url, gpointer data)
{
	GnomeVFSResult error = gnome_vfs_url_show(url);
	if (error != GNOME_VFS_OK) {
		g_print("GnomeVFSResult while trying to launch URL in about dialog: error %u\n", error);
	}
}

static void on_about (GtkWidget *widget, void *data)
{
	const gchar * authors[] = {"Jean Bréfort", NULL};
	const gchar * comments = _("GChem3D is a molecular structures viewer for Gnome");
	/* const gchar * documentors[] = {NULL}; */
	const gchar * copyright = "Copyright \xc2\xa9 2004-2006 Jean Bréfort\n";
	const gchar * license =
		"This program is free software; you can redistribute it and/or\n"
		"modify it under the terms of the GNU General Public License as\n"
		"published by the Free Software Foundation; either version 2 of the\n"
		"License, or (at your option) any later version.\n\n"
		"This program is distributed in the hope that it will be useful,\n"
		"but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
		"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
		"GNU General Public License for more details.\n\n"
		"You should have received a copy of the GNU General Public License\n"
		"along with this program; if not, write to the Free Software\n"
		"Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02111-1307\n"
		"USA";
	
	gtk_about_dialog_set_url_hook(on_about_activate_url, NULL, NULL);

	/* Note to translators: replace the following string with the appropriate credits for you lang */
	const gchar * translator_credits = _("translator_credits");
	gtk_show_about_dialog (NULL,
	                       "name", "GChem3D",
	                       "authors", authors,
	                       "comments", comments,
	                       "copyright", copyright,
	                       "license", license,
	                       "translator_credits", translator_credits,
	                       "version", VERSION,
	                       "website", "http://www.nongnu.org/gchemutils",
	                       NULL);
}

static GtkActionEntry entries[] = {
  { "FileMenu", NULL, N_("_File") },
	  { "Open", GTK_STOCK_OPEN, N_("_Open..."), "<control>O",
		  N_("Open a file"), G_CALLBACK (on_file_open) },
	  { "Print", GTK_STOCK_PRINT, N_("_Print..."), "<control>P",
		  N_("Print the current scene"), G_CALLBACK (on_file_print) },
 	  { "Quit", GTK_STOCK_QUIT, N_("_Quit"), "<control>Q",
		  N_("Quit GChem3D"), G_CALLBACK (on_quit) },
  { "ViewMenu", NULL, N_("_View") },
  { "HelpMenu", NULL, N_("_Help") },
	  { "About", GTK_STOCK_ABOUT, N_("_About"), NULL,
		  N_("About GChem3D"), G_CALLBACK (on_about) }
};

static GtkRadioActionEntry radios[] = {
	{	"BallnStick", NULL, N_("Balls and sticks"), NULL,
		N_("Display a balls and sticks model"),
		0	},
	{	"SpaceFill", "NULL", N_("Space filling"), NULL,
		N_("Display a space filling model"),
		1	},
};

static const char *ui_description =
"<ui>"
"  <menubar name='MainMenu'>"
"    <menu action='FileMenu'>"
"      <menuitem action='Open'/>"
"      <menuitem action='Print'/>"
"      <menuitem action='Quit'/>"
"    </menu>"
"    <menu action='ViewMenu'>"
"      <menuitem action='BallnStick'/>"
"      <menuitem action='SpaceFill'/>"
"	   <separator name='view-sep1'/>"
"      <menuitem action='Background'/>"
"    </menu>"
"    <menu action='HelpMenu'>"
"      <menuitem action='About'/>"
"    </menu>"
"  </menubar>"
"</ui>";

void cb_print_version (const gchar *option_name, const gchar *value, gpointer data, GError **error)
{
	char *version = g_strconcat (_("GChem3d Viewer version: "), VERSION, NULL);
	puts (version);
	g_free (version);
	exit (0);
}

static char *bgcolor = NULL;
static char *display3d = NULL;

static GOptionEntry options[] = 
{
	{ "version", 'v', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, (void*) cb_print_version, N_("Prints GChem3d Viewer version"), NULL },
	{ "bgcolor", 'b', 0, G_OPTION_ARG_STRING, &bgcolor, N_("Background color: white, black or #rrggbb (default is black)"), NULL },
	{ "display3d", 'd', 0, G_OPTION_ARG_STRING, &display3d, N_("How molecules are displayed; possible values are BallnStick (the default) and SpaceFill"), NULL },
	{ NULL }
};

int main(int argc, char *argv[])
{
	GtkWidget *window;
	GtkWidget *viewer, *vbox, *bar;
	GnomeVFSURI *uri, *auri;
	char *path, *dir;
	GtkUIManager *ui_manager;
	GtkActionGroup *action_group;
	GtkAccelGroup *accel_group;
	GError *error = NULL;
	GOptionContext *context;

	textdomain (GETTEXT_PACKAGE);
	gtk_init (&argc, &argv);
	if (!gnome_vfs_init ()) {
		printf ("Could not initialize GnomeVFS\n");
		return 1;
	}

	gcu_element_load_databases ("radii", NULL);
	if (argc > 1 && argv[1][0] == '-') {
		context = g_option_context_new (_(" [file]"));
		g_option_context_add_main_entries (context, options, GETTEXT_PACKAGE);
		g_option_context_add_group (context, gtk_get_option_group (TRUE));
		g_option_context_set_help_enabled (context, TRUE);
		g_option_context_parse (context, &argc, &argv, &error);
		if (error) {
			puts (error->message);
			g_error_free (error);
			return -1;
		}
	}

	if (argc > 1) {
		path = g_get_current_dir ();
		dir = g_strconcat (path, "/", NULL);
		g_free (path);
		uri = gnome_vfs_uri_new (dir);
		auri = gnome_vfs_uri_resolve_relative (uri, argv[1]);
		path = gnome_vfs_uri_to_string (auri, GNOME_VFS_URI_HIDE_NONE);
		viewer = gtk_chem3d_viewer_new (path);
		g_free (path);
		gnome_vfs_uri_unref (auri);
		gnome_vfs_uri_unref (uri);
		g_free (dir);
	} else
		viewer = gtk_chem3d_viewer_new("");

	/* Initialize libgoffice */
	libgoffice_init ();

	window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title(GTK_WINDOW(window), "GtkChem3dViewer");
	gtk_window_set_default_size (GTK_WINDOW(window), 200, 230);
	g_signal_connect(GTK_OBJECT(window), "destroy",
		GTK_SIGNAL_FUNC(gtk_main_quit),
		NULL);
	vbox = gtk_vbox_new (FALSE, 0);
	gtk_container_add (GTK_CONTAINER (window), vbox);
	ui_manager = gtk_ui_manager_new ();
	action_group = gtk_action_group_new ("MenuActions");
	gtk_action_group_set_translation_domain (action_group, GETTEXT_PACKAGE);
	gtk_action_group_add_actions (action_group, entries, G_N_ELEMENTS (entries), window);
	GOActionComboColor *combo = go_action_combo_color_new ("Background", "gcu_Background", "", RGBA_BLACK, NULL);
	g_object_set (G_OBJECT (combo), "label", _("Background color"), "tooltip",
		_("Choose a new background color"), NULL);
	g_signal_connect (G_OBJECT (combo), "activate", G_CALLBACK (on_color_changed), viewer);
	gtk_action_group_add_action (action_group, GTK_ACTION (combo));
	gtk_action_group_add_radio_actions (action_group, radios, G_N_ELEMENTS (radios), 0, G_CALLBACK (on_display), viewer);
	gtk_ui_manager_insert_action_group (ui_manager, action_group, 0);
	accel_group = gtk_ui_manager_get_accel_group (ui_manager);
	gtk_window_add_accel_group (GTK_WINDOW (window), accel_group);
	error = NULL;
	if (!gtk_ui_manager_add_ui_from_string (ui_manager, ui_description, -1, &error)) {
		g_message ("building menus failed: %s", error->message);
		g_error_free (error);
		exit (EXIT_FAILURE);
	}
	bar = gtk_ui_manager_get_widget (ui_manager, "/MainMenu");
	gtk_box_pack_start (GTK_BOX (vbox), bar, FALSE, FALSE, 0);
	gtk_container_add(GTK_CONTAINER(vbox), viewer);
	g_object_set_data (G_OBJECT (window), "viewer", viewer);
	gtk_widget_show_all(window);
	if (bgcolor) {
		g_object_set (G_OBJECT (viewer), "bgcolor", bgcolor, NULL);
		g_free (bgcolor);
	}

	if (display3d) {
		Display3DMode mode = BALL_AND_STICK;
		if (strcmp (display3d, "SpaceFill"))
			mode = SPACEFILL;
		else if (strcmp (display3d, "BallnStick"))
			g_warning (_("Unknown display mode"));
		g_object_set (G_OBJECT (viewer), "display3d", mode, NULL);
		g_free (display3d);
	}

	gtk_main();
	
	return(0);
}
